### PR TITLE
Use retries for http calls

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/extensions/c8y_http_proxy/src/actor.rs [hunk 1]
```diff
@@ -43,10 +51,35 @@ use tedge_http_ext::HttpResult;
...
+        }
+    }
+}
+
 pub struct C8YHttpProxyActor {
     end_point: C8yEndPoint,
     child_devices: HashMap<String, String>,
     peers: C8YHttpProxyMessageBox,
+    token: Option<String>,
+    id_usage: InternalIDUsage,
```
Comment (didier-wenzek) [inline crates/extensions/c8y_http_proxy/src/actor.rs:82]: - Why do you need to know how the internal id is used by a request?
- This seems weird to have this usage attached to the actor while this depends on the request.
- I got a quick look to the code where this used. Please stop doing this. Use method parameters to change the behavior of method!

**Review by didier-wenzek — COMMENTED** [review]
This is too complex.

I failed to wrap my head around the sequence `execute_retry` -> `build_request_using_parts_and_fresh_jwt_token` -> `exec_to_get_internal_id` -> `try_get_internal_id` -> `get_c8y_internal_child_id` -> `retry_once_with_fresh_jwt_token` -> `retry_once_with_fresh_jwt_token` ....

You don't need method name *retry_once* to retry something once.

I would use something akin to:
- methods that are given all the required tokens (internal id and jwt) and simply build and exec the http request.
- methods that use cached tokens to call a first time the former method and depending of the error invalid and renew the jwt token (on 402 and 403) or the internal id (on 404) and retry *once* by calling the same method with the fresh tokens.
- methods

File: crates/extensions/c8y_http_proxy/src/actor.rs (hunk 1 above)
Comment (PradeepKiruvale) [reply crates/extensions/c8y_http_proxy/src/actor.rs:82]: Refactored the code

- Moved the token to the `end_point` struct
- Using the `method`. to update the `body` or `URL` of the request.

File: crates/extensions/c8y_http_proxy/src/actor.rs [hunk 2]
```diff
@@ -43,10 +51,35 @@ use tedge_http_ext::HttpResult;
...
+    Url,
+    Body,
+}
+
+struct HttpRequestParts {
+    method: Method,
+    uri: Uri,
+    body: String,
+    headers: HeaderMap,
+}
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/extensions/c8y_http_proxy/src/actor.rs:64]: I use this one to keep all the components of the `HTTP request` to rebuild the `request` and retry the call with a fresh `jwt` token and `internal-id` if the request fails.

**Pull request closed without merging**